wgengine/packet/ip.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package packet
+
+import (
+	"fmt"
+	"net"
+)
+
+// IP is an IPv4 address.
+type IP uint32
+
+// NewIP converts a standard library IP address into an IP.
+// It panics if b is not an IPv4 address.
+func NewIP(b net.IP) IP {
+	b4 := b.To4()
+	if b4 == nil {
+		panic(fmt.Sprintf("To4(%v) failed", b))
+	}
+	return IP(get32(b4))
+}
+
+func (ip IP) String() string {
+	return fmt.Sprintf("%d.%d.%d.%d", byte(ip>>24), byte(ip>>16), byte(ip>>8), byte(ip))
+}
+
+// IPProto encodes an IP protocol, such as TCP or UDP.
+type IPProto uint8
+
+// IPProto is either a real IP protocol (ICMP, TCP, UDP, ...) or an flag like Junk.
+// If it is a real IP protocol, its value corresponds to its IP protocol number.
+// TODO(dmytro): flags should be taken out of here.
+const (
+	// Junk is deliberately the zero value of IPProto.
```

Comment:
// Junk represents an unknown or supported protocol; it's deliberately the zero value.